### PR TITLE
fix issue #1216

### PR DESCRIPTION
**Pull request by rikuayanokozy**
fix issue [#1216](https://github.com/google/closure-compiler/issues/1216)

### CODE REVIEW

File: test/com/google/javascript/jscomp/parsing/TypeSyntaxTest.java [hunk 1]
```diff
@@ -111,6 +111,16 @@ public void testTypeInDocAndSyntax() {
     parse("var /** string */ foo: string = 'hello';");
   }
 
+  public void testTypedGetterSetterDeclaration() {
+    Node n = parse("var x = {get a(): number {\n}};", LanguageMode.ECMASCRIPT6_TYPED);
+    assertDeclaredType("number type", numberType(),
+        n.getFirstChild().getFirstChild().getFirstChild().getFirstChild().getFirstChild());
+    n = parse("var x = {set a(v: number) {\n}};", LanguageMode.ECMASCRIPT6_TYPED);
+    assertDeclaredType("number type", numberType(),
+        n.getFirstChild().getFirstChild().getFirstChild().getFirstChild().getFirstChild().getChildAtIndex(1)
```
Comment (blickly) [inline test/com/google/javascript/jscomp/parsing/TypeSyntaxTest.java]: Can we add one more test that we get a parse error for setters with a return?
e.g.
  var x = {set a(x): number {}}

Also, if it does work, consider adding a test to NewTypeInferenceES6TypedTest.java, such as the following:
    typeCheck(
        "var x = {get a(): number { return 'str'; }}",
        NewTypeInference.RETURN_NONDECLARED_TYPE);

(I consider this a more straightforward way of checking that the types are correctly parsed than traversing the AST)

Comment (mprobst) [inline test/com/google/javascript/jscomp/parsing/TypeSyntaxTest.java]: nit: I think there were some test utilities that allowed you to search a tree for a node of a particular type - that might be more readable than this tree navigation style.

**Comment by mprobst** [comment]
Looks very good to me.

**Comment by alexeagle** [comment]
Hmm, that's likely. We need a round-trip test in `test/com/google/javascript/jscomp/Es6TypedIntegrationTest.java`

I added
```
  public void testClassGetterSetter() throws Exception {
    test(createCompilerOptions(),
        "class C {" +
                " get x(): number {return 1;}" +
                " set x(n: number): void {} }", "");
  }
```
but since it fails to parse, I'm not sure if we will re-emit correctly.

**Comment by rikuayanokozy** [comment]
@blickly 
I added a check to parse setter and it will throw an error when setters have returns declaration.
Two test case works now.

**Comment by mprobst** [comment]
@rikuayanokozy could you add @alexeagle's suggested test?

**Comment by rikuayanokozy** [comment]
I added Object.defineProperties to IntegrationTestCase and the test can be passed.
But output JSDoc is a bit strange:
```
Object.defineProperties(function(){}.prototype,{x:{configurable:!0,enumerable:!0,/**@return {number} @this {C} */get:/**@return {number} */function(){return 1},/**@this {C} */set:function(){}}})
```

**Comment by mprobst** [comment]
That looks as if it tries to down-compile the getters and setters to `defineProperty`, which is correct as far as your code is concerned.

But then does not strip the comments, which it really should. @blickly any idea?

**Comment by rikuayanokozy** [comment]
https://github.com/google/closure-compiler/issues/1158#issuecomment-148512603
I rewrote the getter/setter down-compile in Es6ToEs3Converter to provide better code size, but it still have bug on JSDoc generate.
Maybe we can make change based on this?

**Comment by blickly** [comment]
I thought the non-transpiled parse-and-print test cases were in CodePrinterEs6TypedTest.java

Since this fix is a parser fix, it seems to me like checking the transpiled output (like in IntegrationTest) is out of scope for this PR.

**Comment by rikuayanokozy** [comment]
Agree.
Should we merge this PR then open a new issue for that problem?

I thought the non-transpiled parse-and-print test cases were in
CodePrinterEs6TypedTest.java

Since this fix is a parser fix, it seems to me like checking the transpiled
output (like in IntegrationTest) is out of scope for this PR.

—
Reply to this email directly or view it on GitHub
<https://github.com/google/closure-compiler/pull/1217#issuecomment-153525396>
.

**Comment by blickly** [comment]
Sorry for the conflicting advice, but I'd actually prefer to remove the test case in Es6TypedIntegrationTest.java, since it seems out of scope for this change.

I think that the tests in TypeSyntaxTest.java are already checking parse-and-printing anyway.

**Comment by rikuayanokozy** [comment]
Test case testClassGetterSetter has been removed from Es6TypedIntegrationTest.java.

**Comment by blickly** [comment]
Thanks!

**Pull request merged**